include/pybind11/numpy.h

```diff
@@ -163,11 +216,121 @@ template <typename T> constexpr const int npy_format_descriptor<
 
 #define DECL_FMT(Type, NumPyName, Name) template<> struct npy_format_descriptor<Type> { \
     enum { value = array::API::NumPyName }; \
+    static object dtype() { \
+        if (auto ptr = array::lookup_api().PyArray_DescrFromType_(value)) \
+            return object(ptr, true); \
+        pybind11_fail("Unsupported buffer format!"); \
+    } \
     static PYBIND11_DESCR name() { return _(Name); } }
-DECL_FMT(float, NPY_FLOAT_, "float32"); DECL_FMT(double, NPY_DOUBLE_, "float64"); DECL_FMT(bool, NPY_BOOL_, "bool");
-DECL_FMT(std::complex<float>, NPY_CFLOAT_, "complex64"); DECL_FMT(std::complex<double>, NPY_CDOUBLE_, "complex128");
+DECL_FMT(float, NPY_FLOAT_, "float32");
+DECL_FMT(double, NPY_DOUBLE_, "float64");
+DECL_FMT(bool, NPY_BOOL_, "bool");
+DECL_FMT(std::complex<float>, NPY_CFLOAT_, "complex64");
+DECL_FMT(std::complex<double>, NPY_CDOUBLE_, "complex128");
 #undef DECL_FMT
 
+struct field_descriptor {
+    const char *name;
+    int offset;
+    object descr;
+};
+
+template <typename T>
+struct npy_format_descriptor<T, typename std::enable_if<is_pod_struct<T>::value>::type> {
+    static PYBIND11_DESCR name() { return _("user-defined"); }
```

Comment:
Sure, that makes sense. Just be careful that ``name()`` generally runs at compile time when using C++14 mode, so you can't do any runtime computation.